### PR TITLE
Support Metric level function `serviceRelation` in `MAL` and binding  envoy metrics into the topology.

### PR DESCRIPTION
**Pull request by wankai123**
### Support Metric level function `serviceRelation` in `MAL` and binding envoy metrics into the topology.
- [X] If this is non-trivial feature, paste the links/URLs to the design doc.
- [X] Update the documentation to include this new feature.
- [X] Tests(including UT, IT, E2E) are added to verify the new feature.
- [X] If it's UI related, attach the screenshots below.

- [ ] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #<issue number>.
- [X] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/CHANGES.md).

image

### CODE REVIEW

**Comment by wu-sheng** [comment]
You are a committer now, please choose the milestone, reviewers, and labels on your own.

**Comment by wankai123** [comment]
I'm checking the E2E

File: oap-server/server-bootstrap/src/main/resources/envoy-metrics-rules/envoy.yaml [hunk 1]
```diff
@@ -61,17 +61,17 @@ metricsRules:
     # envoy_cluster_metrics
   - name: cluster_membership_healthy
     exp: envoy_cluster_metrics.tagMatch('metrics_name' , '.+membership_healthy').tagMatch('metrics_name' , 'cluster.outbound.+|cluster.inbound.+').tagNotMatch('cluster_name' , '.+kube-system').sum(['app', 'instance' , 'cluster_name'])
-  - name: cluster_upstream_cx_active
```
Comment (JaredTan95) [inline oap-server/server-bootstrap/src/main/resources/envoy-metrics-rules/envoy.yaml:64]: Whether the metric name can be consistent with the original before?

Comment (wankai123) [reply oap-server/server-bootstrap/src/main/resources/envoy-metrics-rules/envoy.yaml:64]: some prefix +  metric names too long could cause some storage issues. I refactor them and keep consistently

File: oap-server/server-bootstrap/src/main/resources/metadata-service-mapping.yaml [hunk 2]
```diff
@@ -13,5 +13,5 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-serviceName: ${LABELS."service.istio.io/canonical-revision"}.${LABELS."service.istio.io/canonical-name",LABELS."app.kubernetes.io/name",LABELS.app}.${NAMESPACE}
+serviceName: ${LABELS."service.istio.io/canonical-name",LABELS."app.kubernetes.io/name",LABELS.app}
```
Comment (kezhenxu94) [inline oap-server/server-bootstrap/src/main/resources/metadata-service-mapping.yaml:16]: What's the reason to change this back and forth? I remember you change this in the previous PR

File: oap-server/server-bootstrap/src/main/resources/envoy-metrics-rules/envoy-svc-relation.yaml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+#    "P-6H3M"    -- parses as "-6 hours and +3 minutes"
+#    "-P6H3M"    -- parses as "-6 hours and -3 minutes"
+#    "-P-6H+3M"  -- parses as "+6 hours and -3 minutes"
+# </pre>
+
+expSuffix: serviceRelation(DetectPoint.CLIENT, ['app'], ['cluster_name'])
+metricPrefix: envoy_sr
+metricsRules:
+
+
```
Comment (kezhenxu94) [inline oap-server/server-bootstrap/src/main/resources/envoy-metrics-rules/envoy-svc-relation.yaml:36]: Remove these blank lines

File: oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/topology-service-relation.yml [hunk 4]
```diff
@@ -120,127 +120,83 @@ templates:
...
-              "queryMetricType":"readMetricsValues",
-              "chartType":"ChartLine",
-              "unit":"PPM - Packets Per Minutes"
-            }
-          ],
-          "client":[
-            {
-              "width":12,
-              "title":"Response Time",
+              "title":"Upstream RQ Increase",
```
Comment (kezhenxu94) [inline oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/topology-service-relation.yml:160]: You already use abbreviations in the metrics names, let's use the full words in the title, otherwise there is nowhere to know the full description, the same for other places

```suggestion
              "title":"Upstream Request Increase",
```

File: oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/topology-service-relation.yml [hunk 5]
```diff
@@ -120,127 +120,83 @@ templates:
             }
           ]
         },
-        "tcp":{
+        "envoy_metrics":{
```
Comment (kezhenxu94) [inline oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/topology-service-relation.yml:123]: Why remove `tcp` template?

File: test/e2e/e2e-test/src/test/java/org/apache/skywalking/e2e/mesh/MetricsServiceE2E.java [hunk 6]
```diff
@@ -124,7 +132,61 @@ void test() throws Exception {
...
+                    final MetricsValueMatcher greaterThanZero = new MetricsValueMatcher();
+                    greaterThanZero.setValue("gt 0");
+                    instanceRespTimeMatcher.setValue(greaterThanZero);
+                    instanceRespTimeMatcher.verify(allValues);
+                }
             }
         }
     }
+
+    /*
```
Comment (kezhenxu94) [inline test/e2e/e2e-test/src/test/java/org/apache/skywalking/e2e/mesh/MetricsServiceE2E.java:164]: Remember to revert this

File: oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/SampleFamily.java [hunk 7]
```diff
@@ -429,10 +448,29 @@ public SampleFamily endpoint(List<String> serviceKeys, List<String> endpointKeys
         return createMeterSamples(new EndpointEntityDescription(serviceKeys, endpointKeys));
     }
 
+    public SampleFamily serviceRelation(DetectPoint detectPoint,
+                                         List<String> serviceKeys,
+                                         List<String> relateServiceKeys) {
```
Comment (kezhenxu94) [inline oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/SampleFamily.java:453]:
```suggestion
    public SampleFamily serviceRelation(DetectPoint detectPoint,
                                        List<String> serviceKeys,
                                        List<String> relateServiceKeys) {
```

File: oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/SampleFamily.java [hunk 8]
```diff
@@ -429,10 +448,29 @@ public SampleFamily endpoint(List<String> serviceKeys, List<String> endpointKeys
...
+        Preconditions.checkArgument(serviceKeys.size() > 0);
+        Preconditions.checkArgument(relateServiceKeys.size() > 0);
+        ExpressionParsingContext.get().ifPresent(ctx -> ctx.scopeType = ScopeType.SERVICE_RELATION);
```
Comment (kezhenxu94) [inline oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/SampleFamily.java:456]:
```suggestion
```

File: oap-server/server-bootstrap/src/main/resources/metadata-service-mapping.yaml (hunk 2 above)
Comment (wankai123) [reply oap-server/server-bootstrap/src/main/resources/metadata-service-mapping.yaml:16]: to keep this consistent with the default ALS `k8sServiceNameRule` by default, and can make the topology consistently if ALS(k8s-mesh) and envoy metrics work at the same time.

File: oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/topology-service-relation.yml (hunk 5 above)
Comment (wankai123) [reply oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/topology-service-relation.yml]: moved back

**Comment by wu-sheng** [comment]
@wankai123 In the previous testing, we can see the number of datasets could cause a huge payload to the backend. We are adding more metrics analysis, so we should set up perf testing for it later.
Could you add an issue to track that we should run load testing for metrics analysis service when we face 100k to 1m pods(envoys) in a single cluster?
@kezhenxu94 Your script should be able to set up this env when we do this, right?

**Review by wu-sheng — APPROVED** [review]
General good to me. @kezhenxu94 Please check the code details.

File: oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/topology-service-relation.yml [hunk 9]
```diff
@@ -120,127 +120,207 @@ templates:
             }
           ]
         },
-        "tcp":{
+         "tcp":{
```
Comment (kezhenxu94) [inline oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/topology-service-relation.yml:123]: these are unnecessary changes and impossible to review, just because of a space

File: test/e2e/e2e-test/src/test/resources/expected/metricsservice/topo.yml [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    name: e2e::reviews
+    type: null
+    isReal: true
+  - id: not null
+    name: e2e::productpage
+    isReal: true
+  - id: not null
+    name: e2e::details
+    type: null
+    isReal: true
```
Comment (kezhenxu94) [inline test/e2e/e2e-test/src/test/resources/expected/metricsservice/topo.yml:27]: Why is there no `ratings` service?

File: docs/en/concepts-and-designs/mal.md [hunk 11]
```diff
@@ -236,14 +236,17 @@ last_server_state_sync_time_in_seconds.tagEqual('production', 'catalog').downsam
...
  - `servcie([svc_label1, svc_label2...])` extracts service level labels from the array argument.
  - `instance([svc_label1, svc_label2...], [ins_label1, ins_label2...])` extracts service level labels from the first array argument, 
                                                                         extracts instance level labels from the second array argument.
  - `endpoint([svc_label1, svc_label2...], [ep_label1, ep_label2...])` extracts service level labels from the first array argument, 
                                                                       extracts endpoint level labels from the second array argument.
-
+ - `serviceRelation(DetectPoint, [svc_label1...], [related_svc_label1...])` DetectPoint including `DetectPoint.CLIENT` and `DetectPoint.SERVER`, 
+   extracts `service` labels from the first array argument, extracts `related service` labels from the second array argument.
+   If `DetectPoint.CLIENT` the `related service` is upstream of the `service`.
+   If `DetectPoint.SERVER` the `related service` is downstream of the `service`.
```
Comment (kezhenxu94) [inline docs/en/concepts-and-designs/mal.md:249]: Should we make the position of `source` and `target` services fixed, i.e. the first array argument being `source service` and the second array argument being `target service`? Because this is a service relation metrics, and the `DetectPoint` should not affect the `source` and `target` services. Also, consider if the `DetectPoint` is a variable (not known constant value), how to determine the position of the 2 labels array arguments?

Comment (wu-sheng) [reply docs/en/concepts-and-designs/mal.md:249]: Agree, I think using the consistent order `[source], [target]` makes more sense, and easier to read.

File: test/e2e/e2e-test/src/test/resources/expected/metricsservice/topo.yml (hunk 10 above)
Comment (wankai123) [reply test/e2e/e2e-test/src/test/resources/expected/metricsservice/topo.yml:27]: Default demo only applied
``` shell
kubectl apply -f https://raw.githubusercontent.com/istio/istio/$ISTIO_VERSION/samples/bookinfo/platform/kube/bookinfo.yaml
kubectl apply -f https://raw.githubusercontent.com/istio/istio/$ISTIO_VERSION/samples/bookinfo/networking/bookinfo-gateway.yaml
```
The routing rule is random

I added：
``` shell 
kubectl apply -f https://raw.githubusercontent.com/istio/istio/$ISTIO_VERSION/samples/bookinfo/networking/destination-rule-all.yaml
kubectl apply -f https://raw.githubusercontent.com/istio/istio/$ISTIO_VERSION/samples/bookinfo/networking/virtual-service-all-v1.yaml
```
in `e2e.istio.yaml` and these rule( reviews:v1) doesn't request ratings.

File: docs/en/concepts-and-designs/mal.md (hunk 11 above)
Comment (wankai123) [reply docs/en/concepts-and-designs/mal.md:249]: `Also, consider if the DetectPoint is a variable (not known constant value), how to determine the position of the 2 labels array arguments?`
Do you mean the `DetectPoint ` might be from metrics labelValue?

File: docs/en/setup/envoy/metrics_service_setting.md [hunk 12]
```diff
@@ -86,6 +86,25 @@ istioctl manifest install -y \
...
 ```
 
+Note:
+`proxyStatsMatcher` only supported by `Istio 1.8+`.
+We recommend using `inclusionRegexps` to reserve the specific metrics which need to analyze that can reduce Memory and CPU overhead.
+For example, OAP used these metrics:
+
+```shell
+istioctl manifest install -y \
+  --set profile=demo `# replace the profile as per your need` \
```
Comment (kezhenxu94) [inline docs/en/setup/envoy/metrics_service_setting.md:96]:
```suggestion
  --set profile=demo # replace the profile as per your need \
```

File: oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/Analyzer.java [hunk 13]
```diff
@@ -240,11 +242,23 @@ private void send(final AcceptableValue<?> v, final long time) {
...
+                    entity.setServiceName(entity.getDestServiceName());
+                    toService(requireNonNull(entity.getDestServiceName()));
+                    serverSide(entity);
+                    break;
+                case CLIENT:
+                    entity.setServiceName(entity.getSourceServiceName());
+                    toService(requireNonNull(entity.getSourceServiceName()));
+                    clientSide(entity);
+                    break;
+            }
```
Comment (kezhenxu94) [inline oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/Analyzer.java:257]: We should generate service traffic for both `source service` and `target service`? And that might be the reason for https://github.com/apache/skywalking/pull/7399/files#r683436684?

**Review by kezhenxu94 — COMMENTED** [review]
Generally LGTM

File: oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/Analyzer.java (hunk 13 above)
Comment (wankai123) [reply oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/Analyzer.java:257]: envoy metric only has client-side DetectPoint, I think the others might only have server-side DetectPoint,
It's not about ` https://github.com/apache/skywalking/pull/7399/files#r683436684`.

**Review by wu-sheng — APPROVED** [review]

File: oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/Analyzer.java (hunk 13 above)
Comment (wu-sheng) [reply oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/Analyzer.java:257]: @kezhenxu94 This seems a logical decision only. Actually all traffic sources of services could be generated by the entity level metrics, rather than relation level.

Comment (wu-sheng) [reply oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/Analyzer.java:257]: But consider in only 1% case, if someone only uses this, but disable another metrics set, the topology and service list could still show up. This is the only purpose.

**Review by kezhenxu94 — APPROVED** [review]
